File: validation/boot/test_boot.py
"""Boot validation: reset -> ROM -> startup -> main."""
from __future__ import annotations

from soc.soc import BOOT_TICKS, ROM_BASE, SRAM_BASE


def test_boot_succeeds(fresh_soc):
    info = fresh_soc.boot(rom_words=[0x12345678, 0x9ABCDEF0])
    assert fresh_soc.cpu.running is True
    assert fresh_soc.cpu.pc == ROM_BASE
    assert fresh_soc.cpu.sp == SRAM_BASE + fresh_soc.config["sram_size_bytes"]
    assert info["boot_ticks"] == BOOT_TICKS == 5
    assert fresh_soc.ticks == BOOT_TICKS
    assert fresh_soc.trace[-1][1] == "main_entry"
    # ROM image landed.
    assert fresh_soc.read(ROM_BASE) == 0x12345678
    # PERF enabled and counting real ticks.
    assert fresh_soc.perf.enabled is True
    assert fresh_soc.perf.cycles == BOOT_TICKS


def test_bss_zeroed_on_boot(soc):
    # Booted fixture zeroes SRAM.
    assert soc.sram_read_word(SRAM_BASE) == 0
    assert soc.sram_read_word(SRAM_BASE + 0x100) == 0
    # Dirty then reboot -> zero again.
    soc.sram_write_word(SRAM_BASE, 0xDEADBEEF)
    assert soc.sram_read_word(SRAM_BASE) == 0xDEADBEEF
    soc.boot()
    assert soc.sram_read_word(SRAM_BASE) == 0


def test_boot_time_deterministic(fresh_soc):
    first = fresh_soc.boot()
    second = fresh_soc.boot()
    assert first["boot_ticks"] == second["boot_ticks"]
    mhz = fresh_soc.config["cpu_frequency_mhz"]
    assert first["boot_ns"] == first["boot_ticks"] * 1000 // mhz

File: soc_c/tests/test_boot.c
/* Boot validation in C: reset -> ROM -> startup -> main.
 * Ports validation/boot/test_boot.py.
 */
#include <assert.h>
#include <stdio.h>

#include "soc.h"

int main(void) {
    soc_t s;
    soc_config_t cfg;
    soc_boot_info_t info;
    uint32_t rom_words[2] = {0x12345678U, 0x9ABCDEF0U};
    uint32_t v = 0;

    soc_config_default(&cfg);
    soc_init(&s, &cfg);
    info = soc_boot(&s, rom_words, 2);
    assert(s.cpu.running == 1);
    assert(s.cpu.pc == SOC_ROM_BASE);
    assert(s.cpu.sp == SOC_SRAM_BASE + s.regions[1].size); /* SP = SRAM top */
    assert(info.boot_ticks == SOC_BOOT_TICKS && info.boot_ticks == 5);
    assert(s.ticks == SOC_BOOT_TICKS);
    assert(soc_read(&s, SOC_ROM_BASE, &v) == SOC_OK && v == 0x12345678U);
    assert(s.perf.enabled == 1);
    assert(s.perf.cycles == SOC_BOOT_TICKS);

    /* BSS zeroed, dirty then reboot zeroes again. */
    assert(soc_sram_read_word(&s, SOC_SRAM_BASE, &v) == SOC_OK && v == 0);
    assert(soc_sram_write_word(&s, SOC_SRAM_BASE, 0xDEADBEEFU) == SOC_OK);
    assert(soc_sram_read_word(&s, SOC_SRAM_BASE, &v) == SOC_OK &&
           v == 0xDEADBEEFU);
    soc_boot(&s, NULL, 0);
    assert(soc_sram_read_word(&s, SOC_SRAM_BASE, &v) == SOC_OK && v == 0);

    /* Determinism + boot_ns derivation. */
    {
        soc_boot_info_t a = soc_boot(&s, NULL, 0);
        soc_boot_info_t b = soc_boot(&s, NULL, 0);
        assert(a.boot_ticks == b.boot_ticks);
        assert(a.boot_ns == a.boot_ticks * 1000 / cfg.cpu_frequency_mhz);
    }
    printf("BOOT OK ticks=%u ns=%u\n", info.boot_ticks, info.boot_ns);
    return 0;
}
